[JIT] greedy by breadth memory planning strategy

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63873
* __->__ #64350
* #64349
* #64348
* #64347
* #62995

Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

Commit:
[JIT] greedy by breadth memory planning strategy

Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769098).

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/memory_planning/greedy_by_breadth.cpp:
```diff
@@ -0,0 +1,102 @@
+      auto size = managed_values[t_val].second;
+      auto mem_alloc = makeAllocation(
+          ulvr, size, ordered_allocations, findOffsetWithSmallestGap);
+      if (allocations.count(mem_alloc.ulvr)) {
```
I'm a bit confused here. `mem_alloc.ulvr` should be equal to `ulvr` by construction in `makeAllocation`. You also have a check above that `allocations.count(ulvr)` do not contain `ulvr`. So how can this `if (allocations.count(mem_alloc.ulvr)) {` ever be true since `makeAllocation` doesn't modify `allocations`?

Krovatkin reviewed (commented):
a few qs

makslevental committed:
Update on "[JIT] greedy by breadth memory planning strategy"


Greedy by operator breadth (a heuristic that allocates largest tensors for operators with largest breadths first) from [this paper](https://arxiv.org/pdf/2001.03288.pdf).

put back breadth test

Differential Revision: [D30769098](https://our.internmc.facebook.com/intern/diff/D30769098)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769098).
